File: src/main/java/de/storchp/opentracks/osmplugin/utils/StatisticElement.java
package de.storchp.opentracks.osmplugin.utils;

import android.content.Context;

import de.storchp.opentracks.osmplugin.R;

public enum StatisticElement {
    CATEGORY() {
        @Override
        public String getText(Context context, TrackStatistics statistics) {
            return context.getString(R.string.stat_category, statistics.getCategory());
        }
    },
    TOTAL_TIME() {
        @Override
        public String getText(Context context, TrackStatistics statistics) {
            return StringUtils.formatElapsedTimeHoursMinutes(context, statistics.getTotalTimeMillis());
        }
    },
    MOVING_TIME() {
        @Override
        public String getText(Context context, TrackStatistics statistics) {
            return StringUtils.formatElapsedTimeHoursMinutes(context, statistics.getMovingTimeMillis());
        }
    },
    DISTANCE_KM() {
        @Override
        public String getText(Context context, TrackStatistics statistics) {
            return StringUtils.formatDistanceInKm(context, statistics.getTotalDistanceMeter());
        }
    },
    DISTANCE_MI() {
        @Override
        public String getText(Context context, TrackStatistics statistics) {
            return StringUtils.formatDistanceInMi(context, statistics.getTotalDistanceMeter());
        }
    },
    SPEED_KM_H() {
        @Override
        public String getText(Context context, TrackStatistics statistics) {
            return StringUtils.formatSpeedInKmPerHour(context, statistics.getAvgMovingSpeedMeterPerSecond());
        }
    },
    PACE_MIN_KM() {
        @Override
        public String getText(Context context, TrackStatistics statistics) {
            return StringUtils.formatPaceMinPerKm(context, statistics.getAvgMovingSpeedMeterPerSecond());
        }
    },
    SPEED_MI_H() {
        @Override
        public String getText(Context context, TrackStatistics statistics) {
            return StringUtils.formatSpeedInMiPerHour(context, statistics.getAvgMovingSpeedMeterPerSecond());
        }
    },
    PACE_MIN_MI() {
        @Override
        public String getText(Context context, TrackStatistics statistics) {
            return StringUtils.formatPaceMinPerMi(context, statistics.getAvgMovingSpeedMeterPerSecond());
        }
    },
    ELEVATION_GAIN_METER() {
        @Override
        public String getText(Context context, TrackStatistics statistics) {
            return StringUtils.formatAltitudeChangeInMeter(context, statistics.getElevationGainMeter());
        }
    },
    ELEVATION_GAIN_FEET() {
        @Override
        public String getText(Context context, TrackStatistics statistics) {
            return StringUtils.formatAltitudeChangeInFeet(context, statistics.getElevationGainMeter());
        }
    };

    public static StatisticElement of(String name) {
        try {
            return valueOf(name);
        } catch (IllegalArgumentException e) {
            return null;
        }
    }

    public abstract String getText(Context context, TrackStatistics statistics);

}

File: src/main/kotlin/de/storchp/opentracks/osmplugin/utils/StatisticElement.kt
package de.storchp.opentracks.osmplugin.utils

import android.content.Context
import de.storchp.opentracks.osmplugin.R
import de.storchp.opentracks.osmplugin.map.TrackStatistics
import java.lang.IllegalArgumentException

enum class StatisticElement {
    CATEGORY {
        override fun getText(context: Context, statistics: TrackStatistics): String? {
            return statistics.category?.let { context.getString(R.string.stat_category, it) }
        }
    },
    TOTAL_TIME {
        override fun getText(context: Context, statistics: TrackStatistics): String? {
            return statistics.totalTime?.let {
                StringUtils.formatElapsedTimeHoursMinutes(context, it)
            }
        }
    },
    MOVING_TIME {
        override fun getText(context: Context, statistics: TrackStatistics): String? {
            return statistics.movingTime?.let {
                StringUtils.formatElapsedTimeHoursMinutes(context, it)
            }
        }
    },
    DISTANCE_KM {
        override fun getText(context: Context, statistics: TrackStatistics): String {
            return StringUtils.formatDistanceInKm(context, statistics.totalDistanceMeter)
        }
    },
    DISTANCE_MI {
        override fun getText(context: Context, statistics: TrackStatistics): String {
            return StringUtils.formatDistanceInMi(context, statistics.totalDistanceMeter)
        }
    },
    SPEED_KM_H {
        override fun getText(context: Context, statistics: TrackStatistics): String? {
            return statistics.avgMovingSpeedMeterPerSecond?.let {
                StringUtils.formatSpeedInKmPerHour(context, it)
            }
        }
    },
    PACE_MIN_KM {
        override fun getText(context: Context, statistics: TrackStatistics): String? {
            return statistics.avgMovingSpeedMeterPerSecond?.let {
                StringUtils.formatPaceMinPerKm(context, it)
            }
        }
    },
    SPEED_MI_H {
        override fun getText(context: Context, statistics: TrackStatistics): String? {
            return statistics.avgMovingSpeedMeterPerSecond?.let {
                StringUtils.formatSpeedInMiPerHour(context, it)
            }
        }
    },
    PACE_MIN_MI {
        override fun getText(context: Context, statistics: TrackStatistics): String? {
            return statistics.avgMovingSpeedMeterPerSecond?.let {
                StringUtils.formatPaceMinPerMi(context, it)
            }
        }
    },
    ELEVATION_GAIN_METER {
        override fun getText(context: Context, statistics: TrackStatistics): String? {
            return statistics.elevationGainMeter?.let {
                StringUtils.formatAltitudeChangeInMeter(context, it)
            }
        }
    },
    ELEVATION_GAIN_FEET {
        override fun getText(context: Context, statistics: TrackStatistics): String? {
            return statistics.elevationGainMeter?.let {
                StringUtils.formatAltitudeChangeInFeet(context, it)
            }
        }
    };

    abstract fun getText(context: Context, statistics: TrackStatistics): String?

}

fun String.toStatisticElement() =
    try {
        StatisticElement.valueOf(this)
    } catch (_: IllegalArgumentException) {
        null
    }